index.js

```diff
@@ -147,43 +167,47 @@ module.exports = async function (url, options) {
-            } catch (error) {
-                clearTimeout(timeoutHandler);
-
-                if (!retry(retryOptions, error, null)) {
-                    if (error.name === 'AbortError') {
-                        return reject(new FetchError(`network timeout at ${url}`, 'request-timeout'));
+            while (!isResponseTimedOut(retryOptions)) {
+                ++attempt;
+                const waitTime = getRetryDelay(retryOptions);
+
+                let timeoutHandler;
+                if (retryOptions.socketTimeout) {
+                    const controller = new AbortController();
+                    timeoutHandler = setTimeout(() => controller.abort(), retryOptions.socketTimeout);
+                    options.signal = controller.signal;
+                }                
+    
+                try {
+                    const response = await fetch(url, options);
+
+                    if (shouldRetry(retryOptions, null, response, waitTime)) {
+                        console.error(`Retrying in ${waitTime} milliseconds, attempt ${attempt} failed (status ${response.status}): ${response.statusText}`);
+                    } else {
+                        // response.timeout should reflect the actual timeout
+                        response.timeout = retryOptions.socketTimeout;
+                        return resolve(response);
                     }
-
-                    return reject(error);
+                } catch (error) {
+                    const response = {status: error.code || 500, statusMessage : error.message || "Unknown server error"};
+                    if (!shouldRetry(retryOptions, error, response, waitTime)) {
+                        if (error.name === 'AbortError') {
+                            return reject(new FetchError(`network timeout at ${url}`, 'request-timeout'));
+                        } else {
+                            return reject(error);
+                        }
+                    }
+                    console.error(`Retrying in ${waitTime} milliseconds, attempt ${attempt} error: ${error.message}`);
+                } finally {
+                    clearTimeout(timeoutHandler);
                 }
-
-                console.error(`Retrying in ${retryOptions.retryInitialDelay} milliseconds, attempt ${attempt - 1} error: ${error.message}`);
+                // Fetch loop is about to repeat, delay as needed first.
+                if (waitTime > 0) {
+                    await new Promise(resolve => setTimeout(resolve, waitTime));
+                }
+                retryOptions.retryInitialDelay *= retryOptions.retryBackoff; // update retry interval
             }
-
-            retryOptions.retryInitialDelay *= retryOptions.retryBackoff; // update retry interval
-            const waitTime = retryDelay(retryOptions);
-            setTimeout(() => { wrappedFetch(); }, waitTime);
+            reject(new FetchError(`network timeout at ${url}`, 'request-timeout'));
```

Comment:
Fun fact, I don't even think there is a valid reason this line will ever be reached unless we have repeated the loop with < 1 ms left and then when we check if time is remaining to start the loop again we have 0ms left.  In reality this should almost never happen but in case it does we should treat it as a timeout.  Unfortunately because this is highly timing dependent, it is nearly impossible to reproduce reliably in a test.